src/Microsoft.ML.TorchSharp/Utils/DefaultDictionary.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.ML.TorchSharp.Utils
+{
+    /// <summary>
+    /// Dictionary with a default value for unseen keys.
+    /// </summary>
+    /// <typeparam name="TKey"></typeparam>
+    /// <typeparam name="TValue"></typeparam>
+    [Serializable]
+    public class DefaultDictionary<TKey, TValue> : IDictionary<TKey, TValue>
+    {
+        private readonly Func<TValue> _init;
+        private readonly Dictionary<TKey, TValue> _dictionary;
+
+        public DefaultDictionary(Func<TValue> init)
+        {
+            _init = init;
+            _dictionary = new Dictionary<TKey, TValue>();
+        }
+
+        public TValue this[TKey key]
+        {
+            get
+            {
+                if (!ContainsKey(key)) Add(key, _init());
+                return _dictionary[key];
+            }
+            set
+            {
+                if (!ContainsKey(key)) Add(key, value);
+                else _dictionary[key] = value;
+            }
+        }
+
+        /*** Below are auto-implemented methods ***/
+
+        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
+        {
+            return _dictionary.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return ((IEnumerable)_dictionary).GetEnumerator();
+        }
+
+        public void Add(KeyValuePair<TKey, TValue> item)
+        {
+            _dictionary.Add(item.Key, item.Value);
+        }
+
+        public void Clear()
+        {
+            _dictionary.Clear();
+        }
+
+        public bool Contains(KeyValuePair<TKey, TValue> item)
+        {
+            return _dictionary.Contains(item);
+        }
+
+        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex)
+        {
+            throw new NotImplementedException();
```

Comment:
>NotImplementedException [](http://example.com/codeflow?start=22&length=23)

can implement this using:

((ICollection<KeyValuePair<TKey, TValue>>)_dic).CopyTo(array, arrayIndex);

Reply:
Awesome. Thanks! Added that in.